README.rst
.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black
.. image:: https://img.shields.io/badge/License-MIT-purple.svg
    :target: https://github.com/Dennis-van-Gils/project-project-trachea-pressure-logger/blob/master/LICENSE.txt

Diffusive bubble growth
=======================
*A Physics of Fluids project.*

A pressure logger at 1 kHz for the human trachea project of Marco de Paoli.

- Github: https://github.com/Dennis-van-Gils/project-trachea-pressure-logger

.. image:: https://github.com/Dennis-van-Gils/project-trachea-pressure-logger/blob/main/images/screenshot.png

Hardware
========
* Adafruit #3857: Adafruit Feather M4 Express - Featuring ATSAMD51 Cortex M4
* MIKROE 4-20 mA R Click (MIKROE-1387): 4-20 mA current loop receiver
* Omega PX2300: Pressure Sensor, differential wet-wet +/- 0.067 bar, current output, 0.25 FS accuracy

Pinout
======

==========        =======
Feather M4        R Click
==========        =======
3V                3.3V
GND               GND
D5                CS
MI                SDO
SCK               SCK
==========        =======

Instructions
============
Download the `latest release <https://github.com/Dennis-van-Gils/project-trachea-pressure-logger/releases/latest>`_
and unpack to a folder onto your drive.

Flashing the firmware
---------------------

Double click the reset button of the Feather while plugged into your PC. This
will mount a drive called `FEATHERBOOT`. Copy
`src_mcu/_build_Feather_M4/CURRENT.UF2 <https://github.com/Dennis-van-Gils/project-trachea-pressure-logger/raw/main/src_mcu/_build_Feather_M4/CURRENT.UF2>`_
onto the Featherboot drive. It will restart automatically with the new
firmware.

Running the application
-----------------------

| Preferred distribution: Anaconda full or Miniconda

    * `Anaconda <https://www.anaconda.com>`_
    * `Miniconda <https://docs.conda.io/en/latest/miniconda.html>`_

Open `Anaconda Prompt` and navigate to the unpacked folder. Run the following to
install the necessary packages: ::

   cd src_python
   conda create -n py39 python=3.9
   conda activate py39
   pip install -r requirements.txt

Now you can run the application.
In Anaconda prompt:

::

    conda activate py39
    ipython main.py


LED status lights
=================

* Solid blue: Booting and setting up
* Solid green: Ready for communication
* Flashing green: Sensor data is being send over USB


CHANGELOG.rst
Changelog
=========

1.0.1 (2022-11-09)
------------------
* Calibrated R Click

1.0.0 (2022-11-09)
------------------
* Initial release


src_python/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""A pressure logger for Marco de Paoli's trachea tube.
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-trachea-pressure-logger"
__date__ = "09-11-2022"
__version__ = "1.0"
print(__url__)
# pylint: disable=bare-except, broad-except, unnecessary-lambda

import os
import sys

# Constants
DAQ_INTERVAL_MS = 1  # [ms] Expected, not ground truth.
CHART_INTERVAL_MS = 10  # [ms]
CHART_HISTORY_TIME = 1800  # [s]

# Global flags
TRY_USING_OPENGL = True
DEBUG = False  # Show debug info in terminal?

# Mechanism to support both PyQt and PySide
# -----------------------------------------

PYQT5 = "PyQt5"
PYQT6 = "PyQt6"
PYSIDE2 = "PySide2"
PYSIDE6 = "PySide6"
QT_LIB_ORDER = [PYQT5, PYSIDE2, PYSIDE6, PYQT6]
QT_LIB = os.getenv("PYQTGRAPH_QT_LIB")

# Parse optional cli argument to enfore a QT_LIB
# cli example: python benchmark.py pyside6
if len(sys.argv) > 1:
    arg1 = str(sys.argv[1]).upper()
    for i, lib in enumerate(QT_LIB_ORDER):
        if arg1 == lib.upper():
            QT_LIB = lib
            break

# pylint: disable=import-error, no-name-in-module, c-extension-no-member
if QT_LIB is None:
    for lib in QT_LIB_ORDER:
        if lib in sys.modules:
            QT_LIB = lib
            break

if QT_LIB is None:
    for lib in QT_LIB_ORDER:
        try:
            __import__(lib)
            QT_LIB = lib
            break
        except ImportError:
            pass

if QT_LIB is None:
    this_file = __file__.split(os.sep)[-1]
    raise Exception(
        f"{this_file} requires PyQt5, PyQt6, PySide2 or PySide6; "
        "none of these packages could be imported."
    )

# fmt: off
if QT_LIB == PYQT5:
    from PyQt5 import QtCore, QtGui, QtWidgets as QtWid    # type: ignore
    from PyQt5.QtCore import pyqtSlot as Slot              # type: ignore
elif QT_LIB == PYQT6:
    from PyQt6 import QtCore, QtGui, QtWidgets as QtWid    # type: ignore
    from PyQt6.QtCore import pyqtSlot as Slot              # type: ignore
elif QT_LIB == PYSIDE2:
    from PySide2 import QtCore, QtGui, QtWidgets as QtWid  # type: ignore
    from PySide2.QtCore import Slot                        # type: ignore
elif QT_LIB == PYSIDE6:
    from PySide6 import QtCore, QtGui, QtWidgets as QtWid  # type: ignore
    from PySide6.QtCore import Slot                        # type: ignore
# fmt: on

QT_VERSION = (
    QtCore.QT_VERSION_STR if QT_LIB in (PYQT5, PYQT6) else QtCore.__version__
)

# pylint: enable=import-error, no-name-in-module, c-extension-no-member
# \end[Mechanism to support both PyQt and PySide]
# -----------------------------------------------

import numpy as np
import pyqtgraph as pg

print(f"{QT_LIB:9s} {QT_VERSION}")
print(f"PyQtGraph {pg.__version__}")

import dvg_pyqt_controls as controls
from dvg_debug_functions import tprint, dprint, print_fancy_traceback as pft
from dvg_pyqt_filelogger import FileLogger
from dvg_pyqtgraph_threadsafe import (
    HistoryChartCurve,
    LegendSelect,
    PlotManager,
)

from dvg_devices.Arduino_protocol_serial import Arduino
from dvg_qdeviceio import QDeviceIO, DAQ_TRIGGER

if TRY_USING_OPENGL:
    try:
        import OpenGL.GL as gl  # pylint: disable=unused-import
        from OpenGL.version import __version__ as gl_version
    except:
        print("PyOpenGL  not found")
        print("To install: `conda install pyopengl` or `pip install pyopengl`")
    else:
        print(f"PyOpenGL  {gl_version}")
        pg.setConfigOptions(useOpenGL=True)
        pg.setConfigOptions(antialias=True)
        pg.setConfigOptions(enableExperimental=True)
else:
    print("PyOpenGL  disabled")

# Global pyqtgraph configuration
# pg.setConfigOptions(leftButtonPan=False)
pg.setConfigOption("background", controls.COLOR_GRAPH_BG)
pg.setConfigOption("foreground", controls.COLOR_GRAPH_FG)


def get_current_date_time():
    cur_date_time = QtCore.QDateTime.currentDateTime()
    return (
        cur_date_time.toString("dd-MM-yyyy"),  # Date
        cur_date_time.toString("HH:mm:ss"),  # Time
        cur_date_time.toString("yyMMdd_HHmmss"),  # Reverse notation date-time
    )


# ------------------------------------------------------------------------------
#   State
# ------------------------------------------------------------------------------


class State(object):
    """Reflects the actual readings of the Arduino. There should only be one
    instance of the State class.
    """

    def __init__(self):
        self.time = np.nan  # [s]
        self.pres = np.nan  # [mbar]


state = State()

# ------------------------------------------------------------------------------
#   MainWindow
# ------------------------------------------------------------------------------


class MainWindow(QtWid.QWidget):
    def __init__(self, parent=None, **kwargs):
        super().__init__(parent, **kwargs)

        self.setWindowTitle("Trachea pressure logger")
        self.setGeometry(350, 60, 1200, 900)
        self.setStyleSheet(
            controls.SS_TEXTBOX_READ_ONLY
            + controls.SS_GROUP
            + controls.SS_HOVER
        )

        # -------------------------
        #   Top frame
        # -------------------------

        # Left box
        self.qlbl_update_counter = QtWid.QLabel("0")
        self.qlbl_DAQ_rate = QtWid.QLabel("DAQ: nan Hz")
        self.qlbl_DAQ_rate.setStyleSheet("QLabel {min-width: 7em}")
        self.qlbl_recording_time = QtWid.QLabel()

        vbox_left = QtWid.QVBoxLayout()
        vbox_left.addWidget(self.qlbl_update_counter, stretch=0)
        vbox_left.addStretch(1)
        vbox_left.addWidget(self.qlbl_recording_time, stretch=0)
        vbox_left.addWidget(self.qlbl_DAQ_rate, stretch=0)

        # Middle box
        self.qlbl_title = QtWid.QLabel(
            "Trachea pressure logger",
            font=QtGui.QFont("Palatino", 14, weight=QtGui.QFont.Weight.Bold),
        )
        self.qlbl_title.setAlignment(QtCore.Qt.AlignmentFlag.AlignCenter)
        self.qlbl_cur_date_time = QtWid.QLabel("00-00-0000    00:00:00")
        self.qlbl_cur_date_time.setAlignment(
            QtCore.Qt.AlignmentFlag.AlignCenter
        )
        self.qpbt_record = controls.create_Toggle_button(
            "Click to start recording to file"
        )
        self.qpbt_record.clicked.connect(lambda state: logger.record(state))

        vbox_middle = QtWid.QVBoxLayout()
        vbox_middle.addWidget(self.qlbl_title)
        vbox_middle.addWidget(self.qlbl_cur_date_time)
        vbox_middle.addWidget(self.qpbt_record)

        # Right box
        p = {
            "alignment": QtCore.Qt.AlignmentFlag.AlignRight
            | QtCore.Qt.AlignmentFlag.AlignVCenter
        }
        self.qpbt_exit = QtWid.QPushButton("Exit", minimumHeight=30)
        self.qpbt_exit.clicked.connect(self.close)
        self.qlbl_GitHub = QtWid.QLabel(
            f'<a href="{__url__}">GitHub source</a>', **p
        )
        self.qlbl_GitHub.setTextFormat(QtCore.Qt.TextFormat.RichText)
        self.qlbl_GitHub.setTextInteractionFlags(
            QtCore.Qt.TextInteractionFlag.TextBrowserInteraction
        )
        self.qlbl_GitHub.setOpenExternalLinks(True)

        vbox_right = QtWid.QVBoxLayout(spacing=4)
        vbox_right.addWidget(self.qpbt_exit, stretch=0)
        vbox_right.addStretch(1)
        vbox_right.addWidget(QtWid.QLabel(__author__, **p))
        vbox_right.addWidget(self.qlbl_GitHub)
        vbox_right.addWidget(QtWid.QLabel(f"v{__version__}", **p))

        # Round up top frame
        hbox_top = QtWid.QHBoxLayout()
        hbox_top.addLayout(vbox_left, stretch=0)
        hbox_top.addStretch(1)
        hbox_top.addLayout(vbox_middle, stretch=0)
        hbox_top.addStretch(1)
        hbox_top.addLayout(vbox_right, stretch=0)

        # -------------------------
        #   Bottom frame
        # -------------------------

        #  Charts
        # -------------------------

        self.gw = pg.GraphicsLayoutWidget()

        # Plot: Pressure
        p = {
            "color": controls.COLOR_GRAPH_FG.name(),
            "font-size": "12pt",
            "font-family": "Helvetica",
        }
        self.pi_pres = self.gw.addPlot(row=1, col=0)
        self.pi_pres.setLabel("left", text="pressure (mbar)", **p)

        self.plots = [self.pi_pres]
        for plot in self.plots:
            plot.setClipToView(True)
            plot.showGrid(x=1, y=1)
            plot.setLabel("bottom", text="history (s)", **p)
            plot.setMenuEnabled(True)
            plot.enableAutoRange(axis=pg.ViewBox.XAxis, enable=False)
            plot.enableAutoRange(axis=pg.ViewBox.YAxis, enable=True)
            plot.setAutoVisible(y=True)
            plot.setRange(xRange=[-CHART_HISTORY_TIME, 0])

        # Curves
        capacity = round(CHART_HISTORY_TIME * 1e3 / DAQ_INTERVAL_MS)
        PEN_02 = pg.mkPen(color=[252, 15, 192], width=3)

        self.tscurve_pres = HistoryChartCurve(
            capacity=capacity,
            linked_curve=self.pi_pres.plot(pen=PEN_02, name="Pressure"),
        )
        self.tscurves = [self.tscurve_pres]

        #  Group `Readings`
        # -------------------------

        legend = LegendSelect(
            linked_curves=self.tscurves, hide_toggle_button=True
        )

        p = {
            "readOnly": True,
            "alignment": QtCore.Qt.AlignmentFlag.AlignRight,
            "maximumWidth": 54,
        }
        self.qlin_temp = QtWid.QLineEdit(**p)
        self.qlin_pres = QtWid.QLineEdit(**p)

        # fmt: off
        legend.grid.setHorizontalSpacing(6)
        legend.grid.addWidget(self.qlin_pres               , 0, 2)
        legend.grid.addWidget(QtWid.QLabel("± 0.02 mbar"), 0, 3)
        # fmt: on

        qgrp_readings = QtWid.QGroupBox("Readings")
        qgrp_readings.setLayout(legend.grid)

        #  Group 'Log comments'
        # -------------------------

        self.qtxt_comments = QtWid.QTextEdit()
        grid = QtWid.QGridLayout()
        grid.addWidget(self.qtxt_comments, 0, 0)

        qgrp_comments = QtWid.QGroupBox("Log comments")
        qgrp_comments.setLayout(grid)

        #  Group 'Charts'
        # -------------------------

        self.plot_manager = PlotManager(parent=self)
        self.plot_manager.add_autorange_buttons(linked_plots=self.plots)
        self.plot_manager.add_preset_buttons(
            linked_plots=self.plots,
            linked_curves=self.tscurves,
            presets=[
                {
                    "button_label": "00:30",
                    "x_axis_label": "history (sec)",
                    "x_axis_divisor": 1,
                    "x_axis_range": (-30, 0),
                },
                {
                    "button_label": "01:00",
                    "x_axis_label": "history (sec)",
                    "x_axis_divisor": 1,
                    "x_axis_range": (-60, 0),
                },
                {
                    "button_label": "10:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-10, 0),
                },
                {
                    "button_label": "30:00",
                    "x_axis_label": "history (min)",
                    "x_axis_divisor": 60,
                    "x_axis_range": (-30, 0),
                },
            ],
        )
        self.plot_manager.add_clear_button(linked_curves=self.tscurves)
        self.plot_manager.perform_preset(1)

        qgrp_chart = QtWid.QGroupBox("Charts")
        qgrp_chart.setLayout(self.plot_manager.grid)

        vbox = QtWid.QVBoxLayout()
        vbox.addWidget(qgrp_readings)
        vbox.addWidget(qgrp_comments)
        vbox.addWidget(qgrp_chart, alignment=QtCore.Qt.AlignmentFlag.AlignLeft)
        vbox.addStretch()

        # Round up bottom frame
        hbox_bot = QtWid.QHBoxLayout()
        hbox_bot.addWidget(self.gw, 1)
        hbox_bot.addLayout(vbox, 0)

        # -------------------------
        #   Round up full window
        # -------------------------

        vbox = QtWid.QVBoxLayout(self)
        vbox.addLayout(hbox_top, stretch=0)
        vbox.addSpacerItem(QtWid.QSpacerItem(0, 10))
        vbox.addLayout(hbox_bot, stretch=1)

    # --------------------------------------------------------------------------
    #   Handle controls
    # --------------------------------------------------------------------------

    @Slot()
    def update_GUI(self):
        str_cur_date, str_cur_time, _ = get_current_date_time()
        self.qlbl_cur_date_time.setText(f"{str_cur_date}    {str_cur_time}")
        self.qlbl_update_counter.setText(f"{ard_qdev.update_counter_DAQ:d}")
        self.qlbl_DAQ_rate.setText(
            f"DAQ: {ard_qdev.obtained_DAQ_rate_Hz:.1f} Hz"
        )
        self.qlbl_recording_time.setText(
            f"REC: {logger.pretty_elapsed()}" if logger.is_recording() else ""
        )
        self.qlin_pres.setText(f"{state.pres:.2f}")

    @Slot()
    def update_chart(self):
        if DEBUG:
            tprint("update_chart")

        for tscurve in self.tscurves:
            tscurve.update()


# ------------------------------------------------------------------------------
#   Program termination routines
# ------------------------------------------------------------------------------


def stop_running():
    app.processEvents()
    ard_qdev.quit()
    logger.close()

    print("Stopping timers................ ", end="")
    timer_GUI.stop()
    timer_charts.stop()
    print("done.")


@Slot()
def notify_connection_lost():
    stop_running()

    window.qlbl_title.setText("! ! !    LOST CONNECTION    ! ! !")
    str_cur_date, str_cur_time, _ = get_current_date_time()
    str_msg = f"{str_cur_date} {str_cur_time}\nLost connection to Arduino."
    print(f"\nCRITICAL ERROR @ {str_msg}")
    reply = QtWid.QMessageBox.warning(
        window, "CRITICAL ERROR", str_msg, QtWid.QMessageBox.StandardButton.Ok
    )

    if reply == QtWid.QMessageBox.StandardButton.Ok:
        pass  # Leave the GUI open for read-only inspection by the user


@Slot()
def about_to_quit():
    print("\nAbout to quit")
    stop_running()
    ard.close()


# ------------------------------------------------------------------------------
#   Your Arduino update function
# ------------------------------------------------------------------------------


def DAQ_function():
    # Date-time keeping
    str_cur_date, str_cur_time, str_cur_datetime = get_current_date_time()

    # Listen to the Arduino for sensor readings send out over serial
    success, reply = ard.readline()
    # dprint(reply)

    if not (success):
        dprint(
            "'%s' reports IOError @ %s %s"
            % (ard.name, str_cur_date, str_cur_time)
        )
        return False

    # Parse readings into separate state variables
    try:
        split_reply = (reply.strip()).split("\t")
        state.time = float(split_reply[0]) / 1000
        state.pres = float(split_reply[1])
    except Exception as err:
        pft(err, 3)
        dprint(f"'{ard.name}' reports IOError @ {str_cur_date} {str_cur_time}")
        return False

    # Add readings to chart histories
    window.tscurve_pres.appendData(state.time, state.pres)

    # Logging to file
    logger.update(filepath=str_cur_datetime + ".txt", mode="w")

    # Return success
    return True


# ------------------------------------------------------------------------------
#   File logger
# ------------------------------------------------------------------------------


def write_header_to_log():
    str_cur_date, str_cur_time, _ = get_current_date_time()
    logger.write("[HEADER]\n")
    logger.write(str_cur_date + "\n")
    logger.write(str_cur_time + "\n")
    logger.write(window.qtxt_comments.toPlainText())
    logger.write("\n\n[DATA]\n")
    logger.write("[s]\t[±0.02 mbar]\n")
    logger.write("time\tpres\n")


def write_data_to_log():
    logger.write(f"{logger.elapsed():.3f}\t{state.pres:.2f}\n")


# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":

    # Connect to: Arduino
    ard = Arduino(name="Ard", connect_to_specific_ID="Trachea pressure logger")
    ard.serial_settings["baudrate"] = 115200
    ard.auto_connect(filepath_last_known_port="config/port_Arduino.txt")

    if not (ard.is_alive):
        print("\nCheck connection and try resetting the Arduino.")
        print("Exiting...\n")
        sys.exit(0)

    ard.write("on\n")  # TODO: make nicer

    # Create application and main window
    QtCore.QThread.currentThread().setObjectName("MAIN")  # For DEBUG info
    app = QtWid.QApplication(sys.argv)
    app.setFont(QtGui.QFont("Arial", 9))
    app.aboutToQuit.connect(about_to_quit)
    window = MainWindow()

    # Set up multi-threaded communication: Arduino
    ard_qdev = QDeviceIO(ard)
    ard_qdev.create_worker_DAQ(
        DAQ_trigger=DAQ_TRIGGER.CONTINUOUS,
        DAQ_function=DAQ_function,
        critical_not_alive_count=3,
        debug=DEBUG,
    )
    ard_qdev.signal_DAQ_updated.connect(window.update_GUI)
    ard_qdev.signal_connection_lost.connect(notify_connection_lost)

    # File logger
    logger = FileLogger(
        write_header_function=write_header_to_log,
        write_data_function=write_data_to_log,
    )
    logger.signal_recording_started.connect(
        lambda filepath: window.qpbt_record.setText(
            f"Recording to file: {filepath}"
        )
    )
    logger.signal_recording_stopped.connect(
        lambda: window.qpbt_record.setText("Click to start recording to file")
    )

    # Timers
    timer_GUI = QtCore.QTimer()
    timer_GUI.timeout.connect(window.update_GUI)
    timer_GUI.start(100)

    timer_charts = QtCore.QTimer()
    timer_charts.timeout.connect(window.update_chart)
    timer_charts.start(CHART_INTERVAL_MS)

    # Start threads
    ard_qdev.start()  # NOTE: The DAQ worker thread starts up in a paused state
    ard_qdev.unpause_DAQ()  # Start listening continuously to the Arduino

    # Start the main GUI event loop
    window.show()
    if QT_LIB in (PYQT5, PYSIDE2):
        sys.exit(app.exec_())
    else:
        sys.exit(app.exec())


src_mcu/src/main.cpp
/**
 * @file    main.cpp
 * @author  Dennis van Gils (vangils.dennis@gmail.com)
 * @version https://github.com/Dennis-van-Gils/project-diffusive-bubble-growth
 * @date    09-11-2022
 *
 * @brief   A pressure logger for the Diffusive Bubble Growth setup.
 *
 * @copyright MIT License. See the LICENSE file for details.
 */

/*------------------------------------------------------------------------------
  Description
--------------------------------------------------------------------------------

  Hardware:
  - Adafruit #3857: Adafruit Feather M4 Express - Featuring ATSAMD51 Cortex M4
  - MIKROE 4-20 mA R Click (MIKROE-1387): 4-20 mA current loop receiver
  - RS PRO #797-5018: Pressure Sensor, 0-10 bar, current output

  Reports the readings of the pressure sensor over the serial port as follows:
  >> [millis timestamp] \t [averaged millibars] \n

  Pinout:
    Feather M4        R Click
    ----------        -------
    3V                3.3V
    GND               GND
    D5                CS
    MI                SDO
    SCK               SCK

  The NeoPixel RGB LED of the Feather M4 will indicate its status:
  - Blue : We're setting up
  - Green: Running okay
  - Every read out, the LED will flash brightly in green
------------------------------------------------------------------------------*/

#include "Adafruit_NeoPixel.h"
#include "DvG_StreamCommand.h"
#include "MIKROE_4_20mA_RT_Click.h"

#include <Arduino.h>

// Instantiate serial port listener for receiving ASCII commands
const uint8_t CMD_BUF_LEN = 16;  // Length of the ASCII command buffer
char cmd_buf[CMD_BUF_LEN]{'\0'}; // The ASCII command buffer
DvG_StreamCommand sc(Serial, cmd_buf, CMD_BUF_LEN);

// Onboard NeoPixel
Adafruit_NeoPixel neo(1, PIN_NEOPIXEL, NEO_GRB + NEO_KHZ800);
const uint8_t NEO_DIM = 2;    // Brightness level for dim intensity [0 -255]
const uint8_t NEO_BRIGHT = 6; // Brightness level for bright intensity [0 - 255]
const uint32_t FLASH_LENGTH = 100; // [us]

/*------------------------------------------------------------------------------
  MIKROE 4-20 mA R click board for reading out the pressure sensor
------------------------------------------------------------------------------*/

// Cable select pin
const uint8_t PIN_R_CLICK = 5;

// Calibrated against a multimeter @ 11-10-2022 by DPM van Gils
const RT_Click_Calibration R_CLICK_CALIB{4.11, 20.02, 830, 4002};

// Single R click readings fluctuate a lot and so we will employ an exponential
// moving average (EMA) by using oversampling and subsequent low-pass filtering
const uint32_t EMA_DT = 100; // Desired oversampling interval [µs]
const float EMA_LP = 1000;   // Low-pass filter cut-off frequency [Hz]

R_Click R_click(PIN_R_CLICK, R_CLICK_CALIB, EMA_DT, EMA_LP);

/*------------------------------------------------------------------------------
  PR PRO pressure sensor, type 797-5018
------------------------------------------------------------------------------*/

// Structure to hold the RS PRO pressure sensor calibration parameters.
// The parameters are found on the calibration sheet supplied with the sensor.
struct Pressure_Calibration {
  float zero_mA;
  float span_mA;
  float full_range_bar;
};

// Calibration parameters: ESTIMATED, no calibration sheet supplied with sensor
// NOTE: The sensor is DIFFERENTIAL, so +/- 0.0689, hence the full range is
// double this value
const Pressure_Calibration PRESSURE_CALIB{4.0, 16, 0.0689 * 2.0f};

// NOTE: The sensor is DIFFERENTIAL, so we have to add 12 mA to get to 0 bar
inline float mA2bar(float mA, const Pressure_Calibration calib) {
  return (mA - calib.zero_mA) / calib.span_mA * calib.full_range_bar -
         calib.full_range_bar / 2.f;
}

/*------------------------------------------------------------------------------
  Readings
------------------------------------------------------------------------------*/

struct Readings {
  float pres_bitval = NAN; // Pressure sensor [EMA bitval]
  float pres_mA = NAN;     // Pressure sensor [mA]
  float pres_bar = NAN;    // Pressure sensor [bar]
};
Readings readings; // Structure holding the sensor readings

/*------------------------------------------------------------------------------
  setup
------------------------------------------------------------------------------*/

void setup() {
  neo.begin();
  neo.setPixelColor(0, neo.Color(0, 0, NEO_BRIGHT)); // Blue: We're in setup()
  neo.show();

  pinMode(PIN_LED, OUTPUT);
  digitalWrite(PIN_LED, LOW);

  Serial.begin(9600);
  R_click.begin();

  neo.setPixelColor(0, neo.Color(0, NEO_DIM, 0)); // Green: All set up
  neo.show();
}

/*------------------------------------------------------------------------------
  loop
------------------------------------------------------------------------------*/

void loop() {
  uint32_t now = millis();
  static uint32_t tick = now;
  static bool flash = false;
  static uint32_t flash_tick = now;
  static bool report = false;

  // Update R-click readings
  R_click.poll_EMA();

  // Listen for incoming commands over serial
  if (sc.available()) {
    char *str_cmd = sc.getCommand();

    // Flash the NeoPixel bright green to indicate we received a command
    neo.setPixelColor(0, neo.Color(0, NEO_BRIGHT, 0));
    neo.show();
    flash = true;
    flash_tick = now;

    if (strcmp(str_cmd, "id?") == 0) {
      // Report identity string
      report = false;
      Serial.println("Arduino, Trachea pressure logger");

    } else if (strcmp(str_cmd, "on") == 0) {
      report = true;

    } else if (strcmp(str_cmd, "off") == 0) {
      report = false;
    }
  }

  if (report && (tick - now >= 1)) {
    tick = now;
    readings.pres_mA = R_click.get_EMA_mA();
    readings.pres_bar = mA2bar(readings.pres_mA, PRESSURE_CALIB);
    Serial.print(now);
    // Serial.write('\t');
    // Serial.print(readings.pres_mA, 2);
    Serial.write('\t');
    Serial.println(readings.pres_bar * 1000, 2); // Reports in mbar
  }

  if (flash && (now - flash_tick >= FLASH_LENGTH)) {
    // Set the NeoPixel back to dim green to indicate we're in an idle state
    flash = false;
    neo.setPixelColor(0, neo.Color(0, NEO_DIM, 0));
    neo.show();
  }
}


src_python/inspect_all_logs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Scan for all log files in the current folder. Those that are missing a plot
figure will be processed. Useful tool for quick inspection.
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-trachea-pressure-logger"
__date__ = "09-11-2022"
__version__ = "1.0"

import os
import re
from LogInspector import read_log, plot_log

if __name__ == "__main__":
    my_path = os.getcwd()
    file_list = [
        f
        for f in os.listdir(my_path)
        if os.path.isfile(os.path.join(my_path, f))
    ]

    for filename in file_list:
        # Look for files matching: ######_###### [+any extra chars] .txt
        p = re.compile("\d{6}_\d{6}(.*?)\.(txt|TXT)$")
        if p.match(filename):
            # Found a matching file
            # Now check if the same filename exists ending with .png
            filename_png = filename[0:-4] + ".png"

            if not os.path.isfile(filename_png):
                # Figure does not yet exists. Create.
                print(f"Reading file: {filename}")
                log = read_log(filename)
                plot_log(log)


src_python/LogInspector.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Provides functions for reading in and plotting log files specific to the
current project.

When called straight from the command line it will open a file-navigator.
"""
__author__ = "Dennis van Gils"
__authoremail__ = "vangils.dennis@gmail.com"
__url__ = "https://github.com/Dennis-van-Gils/project-trachea-pressure-growth"
__date__ = "09-11-2022"
__version__ = "1.0"


from pathlib import Path
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt


class Log:
    """Structure that holds the timeseries data and additional information of a
    log file specific to the current project."""

    def __init__(self):
        self.filename = ""
        self.header_date = ""
        self.header_time = ""
        self.header_msg = [""]
        self.time = np.array([])
        self.pres = np.array([])


# ------------------------------------------------------------------------------
#   read_log
# ------------------------------------------------------------------------------


def read_log(filepath=None) -> Log:
    """Reads in a log file.

    Args:
        filepath (pathlib.Path, str):
            Path to the data file to open.

    Returns: instance of Log class
    """
    if isinstance(filepath, str):
        filepath = Path(filepath)

    if not isinstance(filepath, Path):
        raise Exception(
            "Wrong type passed to read_log(). "
            "Should be (str) or (pathlib.Path)."
        )

    if not filepath.is_file():
        raise Exception("File can not be found\n %s" % filepath.name)

    with filepath.open() as f:
        log = Log()

        # Scan the first lines for the start of the header and data sections
        MAX_LINES = 100  # Stop scanning after this number of lines
        str_header = []
        success = False
        for i_line in range(MAX_LINES):
            str_line = f.readline().strip()

            if str_line.upper() == "[HEADER]":
                # Simply skip
                pass
            elif str_line.upper() == "[DATA]":
                # Found data section. Exit loop.
                i_line_data = i_line
                success = True
                break
            else:
                # We must be in the header section now
                str_header.append(str_line)

        if not success:
            raise Exception(
                "Incorrect file format. Could not find [DATA] section."
            )

        # Read in all data columns including column names
        tmp_table = np.genfromtxt(
            filepath.name,
            delimiter="\t",
            names=True,
            skip_header=i_line_data + 2,
        )

        # Rebuild into a Matlab style 'struct'
        log.filename = filepath.name[0:-4]
        log.header_date = str_header[0]
        log.header_time = str_header[1]
        log.header_msg = str_header[2:]
        log.time = tmp_table["time"]
        log.pres = tmp_table["pres"]

    return log


# ------------------------------------------------------------------------------
#   plot_log
# ------------------------------------------------------------------------------

# Special characters
CHAR_PM = "\u00B1"
CHAR_DEG = "\u00B0"

# Colors
cm = (
    np.array(
        [
            [255, 255, 0],
            [252, 15, 192],
            [0, 255, 255],
            [255, 255, 255],
            [255, 127, 39],
            [0, 255, 0],
        ]
    )
    / 255
)


def plot_log(log: Log):
    # Lay-out
    mpl.style.use("dark_background")
    mpl.rcParams["font.size"] = 12
    # mpl.rcParams['font.weight'] = "bold"
    mpl.rcParams["axes.titlesize"] = 14
    mpl.rcParams["axes.labelsize"] = 14
    mpl.rcParams["axes.titleweight"] = "bold"
    mpl.rcParams["axes.formatter.useoffset"] = False
    # mpl.rcParams["axes.labelweight"] = "bold"
    mpl.rcParams["lines.linewidth"] = 2
    mpl.rcParams["grid.color"] = "0.25"

    fig1 = plt.figure(figsize=(16, 10), dpi=90)
    fig1.canvas.manager.set_window_title(log.filename)

    ax1 = fig1.add_subplot(1, 1, 1)

    # Plot pressure
    ax1.plot(log.time, log.pres, color=cm[1], label="Pressure")
    ax1.set_title(f"Pressure ({CHAR_PM} 0.02 mbar)")
    ax1.set_xlabel("time (s)")
    ax1.set_ylabel("pressure (mbar)")
    ax1.grid(True)

    # Finalize lay-out
    ax_w = 0.9
    ax_h = 0.8
    ax1.set_position([0.08, 0.06, ax_w, ax_h])

    # Save figure
    img_format = "png"
    fn_save = f"{log.filename}.{img_format}"
    plt.savefig(
        fn_save,
        dpi=90,
        orientation="portrait",
        format=img_format,
        transparent=False,
    )
    print(f"Saved image: {fn_save}")


# ------------------------------------------------------------------------------
#   Main
# ------------------------------------------------------------------------------

if __name__ == "__main__":
    import os
    import sys
    import tkinter
    from tkinter import filedialog

    # Check for optional input arguments
    filename_supplied = False
    for arg in sys.argv[1:]:
        filename = arg
        filename_supplied = True

    root = tkinter.Tk()
    root.withdraw()  # Disable root window

    if not filename_supplied:
        filename = filedialog.askopenfilename(
            initialdir=os.getcwd(),
            title="Select data file",
            filetypes=(("text files", "*.txt"), ("all files", "*.*")),
        )
    if filename == "":
        sys.exit(0)

    root.destroy()  # Close file dialog

    print(f"Reading file: {filename}")
    plot_log(read_log(filename))
    plt.show()
